### ISSUE TITLE
[Q] You must call wandb.init() before wandb.log(): error, how to resolve this?

### ISSUE THREAD

**Issue by qm-intel** [wandb/wandb#6556]
Dear @thanos-wandb @vanpelt @lukas 
After few training steps, I am getting the following error, and I am not sure what is the reason: 

```
Training 10:   5%|▌         | 10/185 [00:10<02:15,  1.29it/s].Error  File "/home/users/se/.conda/envs/torch-env/lib/python3.8/site-packages/wandb/sdk/lib/preinit.py", line 36, in preinit_wrapper
wandb.errors.Errorwandb.errors[A: You must call wandb.init() before wandb.log()
    : You must call wandb.init() before wandb.log()
.Error: You must call wandb.init() before wandb.log()
raise wandb.Error(f"You must call wandb.init() before {name}()")
wandb.errors.Error: You must call wandb.init() before wandb.log()
```

My distributed setting is like this:

```
# Set the device
    gpu = int(os.environ['LOCAL_RANK'])
    device = torch.device(f'cuda:{gpu}')
    torch.cuda.set_device(device)
....

model = distributed_training_init(model)

...
    if dist.get_rank() == 0:
        status = "online" #if wandb_online else "offline"
        run = wandb.init( project=project_name, name=experiment_name, mode=status)
        config.run_id = run.id
        run_dir = "/".join(run.dir.split('/')[:-1])
        config.run_dir = run_dir
        wandb.config.update(config)
        run.watch(model)
        with open(os.path.join(run_dir, "experiment_config.yaml"), 'w') as outfile:
            yaml.dump(config_hyperparam, outfile, default_flow_style=False)

        """wandb.config.update(config)"""

        # os.mkdir(f"models/{config.run_id}")  #-
        os.mkdir(f"{config.run_dir}/models")

...
```
**When I remove the ` if dist.get_rank() == 0:` wandb creates 4 run folder if I have 4 GPUs.** 

and this is  the function for 

```
def distributed_training_init(model, backend='nccl', sync_bn=False):
    if sync_bn:
        model = torch.nn.SyncBatchNorm.convert_sync_batchnorm(model)

    rank = int(os.environ['RANK'])
    world_size = int(os.environ['WORLD_SIZE'])
    gpu = int(os.environ['LOCAL_RANK'])

    print(rank, world_size, gpu)
    torch.distributed.init_process_group(backend, world_size=world_size,
                                         rank=rank, init_method='env://')

    print('gpu', gpu)
    model = torch.nn.parallel.DistributedDataParallel(model, device_ids=[gpu], output_device=gpu,find_unused_parameters=True)
    return model
```

How to resolve this issue?

**Comment by thanos-wandb** [comment]
Hi @qm-intel, thank you for reporting this issue. It appears that the error is likely related to your multiprocessing setup, where some non-zero ranks are attempting to log to W&B without knowledge of the run's singleton. 

It seems you are trying to log from a single process in your distributed training approach. In this scenario, all W&B calls should be made from rank zero, while the rest should be no-ops. For your convenience, we have provided guidelines on [this approach in our Docs page](https://docs.wandb.ai/guides/track/log/distributed-training#method-1-one-process), which you may find useful. Additionally, you can refer to [this example in our Github repository](https://github.com/wandb/examples/blob/master/examples/pytorch/pytorch-ddp/log-ddp.py) that demonstrates this method using PyTorch DDP. 

In order to help us troubleshoot this issue, we would appreciate it if you could provide us with the following information:
- Your current W&B version (wandb --version).
- SDK debug logs: For the affected run, you should check `debug.log` and `debug-internal.log`. These files are under your local folder `wandb/run-<date>_<time>-<run-id>/logs` in the same directory where you’re running your code.
- The 'else' block from the code snippet you shared.
- If there are any other wandb.log() calls in your code, please share those as well. 

This information will greatly assist us in diagnosing the issue.

**Comment by qm-intel** [comment]
@thanos-wandb Thanks for replying to my question and explaining in detail. 

> Additionally, you can refer to [this example in our Github repository](https://github.com/wandb/examples/blob/master/examples/pytorch/pytorch-ddp/log-ddp.py) that demonstrates this method using PyTorch DDP.
Yes, it is really good example and previously I had followed this example and learned about it. But wandb was creating 4 different run folder and saving 4 models considering to allocate 4 GPUs. If I allocate 2 GPUs, it will create two run folder, depending on the number of GPUs. 

To respond your requests:

> Your current W&B version (wandb --version).
wandb, version 0.15.10

> SDK debug logs: For the affected run, you should check debug.log and debug-internal.log. These files are under your local folder wandb/run-<date>_<time>-<run-id>/logs in the same directory where you’re running your code.

I attached to this comment.
[debug.log](https://github.com/wandb/wandb/files/13273319/debug.log)
[debug-internal.log](https://github.com/wandb/wandb/files/13273320/debug-internal.log)
[run.log](https://github.com/wandb/wandb/files/13273321/run.log)

> The 'else' block from the code snippet you shared.
It does not have else block, 

> If there are any other wandb.log() calls in your code, please share those as well.


Here is my code:

  
>      if dist.get_rank() == 0:
>         status = "online" #if wandb_online else "offline"
>         run = wandb.init( project=project_name, name=experiment_name, mode=status)
>         config.run_id = run.id
>         run_dir = "/".join(run.dir.split('/')[:-1])
>         config.run_dir = run_dir
>         wandb.config.update(config)
>         run.watch(model)
>         with open(os.path.join(run_dir, "experiment_config.yaml"), 'w') as outfile:
>             yaml.dump(config_hyperparam, outfile, default_flow_style=False)
>         os.mkdir(f"{config.run_dir}/models")
>
>     step = 0
>     losses = []
>     dur = []
>     training_start_time = time.time()
>     epochs_pbar = tqdm(range(config.epoch))
>     for epoch in epochs_pbar:
>         t0 = time.time()
>         epochs_pbar.set_description(f"Epoch {epoch}")
>         model.train()
> 
>         train_pbar = tqdm(enumerate(dataloader), total=len(dataloader), leave=False)
>         for idx, batch in train_pbar:
>             train_pbar.set_description(f"Training {step:,}")
> 
>             if config.image_size != 64 and config.dataset in ["dfc", "enmap"]:
>                 # select a image_size**2 patch at random location of the tile
>                 x, y = torch.randint(0, 64 - config.image_size, (2,))
>             else:
>                 x, y = 0, 0
>             # batch['img'].shape torch.Size([64, 200, 64, 64])  "b c h w"
>             img = batch["img"][
>                 :, :, x : x + config.image_size, y : y + config.image_size
>             ].to(device)   # shape: torch.Size([64, 200, 8, 8])
> 
>             optimizer.zero_grad()
> 
>             loss = model(img)
> 
>             if torch.isnan(loss):
>                 raise ValueError("Loss is NaN")
> 
>             loss.backward()
>             optimizer.step()
>             step += 1
> 
>             losses.append(loss.detach().item())
> 
>             if step % config.logging_freq == 0:
>                 wandb.log(
>                     {
>                         "epoch": epoch,
>                         "loss": np.array(losses[-1 * config.logging_freq :]).mean(),
>                         "lr": optimizer.param_groups[0]["lr"],
>                     },
>                     step=step,
>                 )
> 
>         # log at end of training epoch (to same step as validation stats below)
>         wandb.log({"epoch": epoch, "loss": loss.item()}, step=step)
> 
>         if epoch % config.model_save_freq == 0:
>             # save model checkpoint along with some statistics
>             stats = {
>                 "losses": torch.tensor(losses),
>                 "config": config,
>                 "model_state_dict": model.state_dict(),
>                 "lr_current": optimizer.param_groups[0]["lr"],
>                 "input": img.detach(),
>                 "transformer_input": img,
>             }
>             torch.save(
>                 stats,
>                 f"{config.run_dir}/models/model_{config.encoder_name}_ep{epoch}.pth",
>             )   # changed {config.run_id} to {config.run_dir}
> 
>             if epoch == 10 and config.model_save_freq == 1:
>                 config.model_save_freq = 10
> 
>         if not config.skip_val:
>             # validation
>             with torch.no_grad():
>                 val_losses = []
>                 val_accs = []
>                 model.eval()
>                 val_pbar = tqdm(
>                     enumerate(val_dataloader), total=len(val_dataloader), leave=False
>                 )
>                 for idx, batch in val_pbar:
>                     val_pbar.set_description(f"Validation {step:,}")
>                     img_whole = batch["img"]
> 
>                     if config.image_size != 64 and config.dataset in ["dfc", "enmap"]:
>                         # sliding window with stride == window size
>                         for x in range(0, 64, config.image_size):
>                             for y in range(0, 64, config.image_size):
>                                 img = img_whole[
>                                     :,
>                                     :,
>                                     x : x + config.image_size,
>                                     y : y + config.image_size,
>                                 ].to(device)
> 
>                                 loss = model(img)
> 
>                                 val_losses.append(loss.detach().item())
> 
>                     else:
>                         img = img_whole.to(device)
>                         loss = model(img)
>                         val_losses.append(loss.detach().item())
> 
>                 wandb.log(
>                     {
>                         "epoch": epoch,
>                         "val_loss": torch.tensor(val_losses).mean().item(),
>                     },
>                     step=step,
>                 )
> 
>             if config.scheduler == "ReduceLROnPlateau":
>                 scheduler.step(torch.tensor(val_losses).mean().item())
>         if config.scheduler == "cosine":
>             scheduler.step(epoch)
>         dur.append(time.time() - t0)
>         print('Epoch: {:03d}, Time(s) {:.4f}'.format(epoch, np.mean(dur)))
>     print('Training finished, took {:.2f}s'.format(time.time() - training_start_time))
>     if dist.get_rank() == 0:
>         wandb.finish()

Thanks for your help

**Comment by qm-intel** [comment]
@thanos-wandb I put the details of my code here https://github.com/wandb/wandb/issues/6556#issuecomment-1796978136

**Comment by thanos-wandb** [comment]
Hi @qm-intel thank you so much for providing all these detailed information. Can you please try to adjust your code so that any call to `wandb.log` is within an `if` statement block similar to below:

```
 if dist.get_rank() == 0:
    wandb.log({...})
```

Would this resolve the issue for you? additionally, I have looked into the debug logs and I can see an error message saying: 
`ERROR   HandlerThread:1648 [system_info.py:_save_conda():86] Error saving conda packages: [Errno 2] No such file or directory: 'conda'`

Have you installed `conda` in the machine where you're running your script?

**Comment by qm-intel** [comment]
@thanos-wandb Thanks for your help.  This solved the issue: 


> Can you please try to adjust your code so that any call to wandb.log is within an if statement block similar to below:


> Would this resolve the issue for you? additionally, I have looked into the debug logs and I can see an error message saying:
> ERROR   HandlerThread:1648 [system_info.py:_save_conda():86] Error saving conda packages: [Errno 2] No such file or directory: 'conda'
> 
> Have you installed conda in the machine where you're running your script?

Since I am running my code on the cluster, I load the anaconda module, and it could solve this error too. Thanks for catching this. I really appreciate your help

**Comment by thanos-wandb** [comment]
Hi @qm-intel I'm glad to hear that the suggested solution resolved your issue! I will close this ticket for now but if you have any further questions or encounter additional challenges, please feel free to reopen the GitHub thread. We'll be more than happy to continue investigating and assisting you. Thanks for your collaboration, and we're here to help whenever you need it.
